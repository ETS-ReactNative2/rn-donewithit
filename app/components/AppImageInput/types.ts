export type AppImageInputProps = {
  onChange?: (values?: string[]) => void
  onAddButtonPress?: () => void
  max?: number
  addButtonBackgroundColor?: string
  addButtonIconName?: string
  addButtonIconColor?: string
  addButtonIconSize?: number
}

export type StateType = {
  values: string[]
}

export enum ActionKind {
  ADD = 'ADD',
  REMOVE = 'REMOVE'
}

export type ActionType = {
  type: ActionKind
  payload: any
}
