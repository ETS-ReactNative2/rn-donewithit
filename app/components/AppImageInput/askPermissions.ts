import * as ImagePicker from 'expo-image-picker'

export const requestMediaPermissions = async () => {
  const { granted } = await ImagePicker.requestMediaLibraryPermissionsAsync()
  if (!granted) alert('You Need to enable permission to access the library!')
}
