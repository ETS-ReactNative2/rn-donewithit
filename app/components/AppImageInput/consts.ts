import COLORS from '../../config/colors'

const CONSTS = {
  PADDING: 10,
  BORDER_RADIUS: 20,
  BOX_SIZE: 100,
  MAX_PHOTOS: 2,
  ADD_BUTTON_ICON_NAME: 'camera',
  ADD_BUTTON_ICON_COLOR: COLORS.GRAY,
  ADD_BUTTON_ICON_SIZE: 38,
  IMAGE_URIS: []
}

export default CONSTS
