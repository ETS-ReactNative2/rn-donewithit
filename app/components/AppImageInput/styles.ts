import { StyleSheet } from 'react-native'
import CONSTS from './consts'

const styles = StyleSheet.create({
  container: {
    flexDirection: 'row',
    flexWrap: 'wrap'
  },
  box: {
    width: CONSTS.BOX_SIZE,
    height: CONSTS.BOX_SIZE,
    borderRadius: CONSTS.BORDER_RADIUS,
    marginRight: CONSTS.PADDING,
    marginTop: CONSTS.PADDING
  },
  image: {},
  addButton: {
    justifyContent: 'center',
    alignItems: 'center'
  }
})

export default styles
