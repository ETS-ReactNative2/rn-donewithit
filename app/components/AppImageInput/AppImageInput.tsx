import React, { useEffect, useReducer } from 'react'
import { View, Image, TouchableOpacity, StyleSheet } from 'react-native'
import AppIcon from '../AppIcon'
import * as ImagePicker from 'expo-image-picker'
import { requestMediaPermissions } from './askPermissions'
import { AppImageInputProps, ActionKind } from './types'
import { reducer } from './reducers'
import COLORS from '../../config/colors'
import CONSTS from './consts'
import styles from './styles'

function AppImageInput({
  onChange,
  max = CONSTS.MAX_PHOTOS,
  addButtonBackgroundColor = COLORS.INPUT_BACKGROUND_COLOR,
  addButtonIconName = CONSTS.ADD_BUTTON_ICON_NAME,
  addButtonIconColor = CONSTS.ADD_BUTTON_ICON_COLOR,
  addButtonIconSize = CONSTS.ADD_BUTTON_ICON_SIZE
}: AppImageInputProps) {
  const customStyles = StyleSheet.create({
    addButton: {
      backgroundColor: addButtonBackgroundColor
    }
  })
  const [state, dispatch] = useReducer(reducer, {
    values: []
  })
  useEffect(() => {
    requestMediaPermissions()
  }, [])
  useEffect(() => {
    typeof onChange === 'function' && onChange(state.values)
  }, [state.values])
  const handleAddButtonPress = async () => {
    try {
      const result = await ImagePicker.launchImageLibraryAsync()
      if (!result.cancelled) {
        dispatch({ type: ActionKind.ADD, payload: result.uri })
      }
    } catch (error) {
      console.log('Error reading an image', error)
    }
  }
  const handleImagePress = (index: number) => {
    dispatch({ type: ActionKind.REMOVE, payload: index })
  }
  return (
    <View style={styles.container}>
      {state.values.map((value: string, index: number) => {
        return (
          <TouchableOpacity
            key={index}
            activeOpacity={0.85}
            onPress={() => {
              handleImagePress(index)
            }}
          >
            <Image
              style={[styles.box, styles.image]}
              source={{
                uri: value
              }}
            />
          </TouchableOpacity>
        )
      })}
      {state.values.length < max && (
        <TouchableOpacity activeOpacity={0.85} onPress={handleAddButtonPress}>
          <View style={[styles.box, styles.addButton, customStyles.addButton]}>
            <AppIcon
              color={addButtonIconColor}
              size={addButtonIconSize}
              name={addButtonIconName}
            />
          </View>
        </TouchableOpacity>
      )}
    </View>
  )
}

export default AppImageInput
