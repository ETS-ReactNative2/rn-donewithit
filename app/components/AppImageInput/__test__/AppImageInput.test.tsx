import React from 'react'
import renderer from 'react-test-renderer'
import AppImageInput from '../index'

jest.mock(
  '@expo/vector-icons/build/vendor/react-native-vector-icons/lib/create-icon-set.js',
  () => {
    return () => ''
  }
)

const props = {
  max: 10
}

describe('AppImageInput', () => {
  it('renders correctly', () => {
    const tree = renderer.create(<AppImageInput {...props} />).toJSON()
    expect(tree).toMatchSnapshot()
  })
})
