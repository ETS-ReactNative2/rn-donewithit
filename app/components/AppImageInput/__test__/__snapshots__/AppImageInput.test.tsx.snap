// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AppImageInput renders correctly 1`] = `
<RCTScrollView
  horizontal={true}
  onContentSizeChange={[Function]}
>
  <View>
    <View
      style={
        Object {
          "flexDirection": "row",
          "flexWrap": "wrap",
        }
      }
    >
      <View
        accessible={true}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "opacity": 1,
          }
        }
      >
        <View
          style={
            Array [
              Object {
                "borderRadius": 20,
                "height": 100,
                "marginRight": 10,
                "marginTop": 10,
                "width": 100,
              },
              Object {
                "alignItems": "center",
                "justifyContent": "center",
              },
              Object {
                "backgroundColor": "#f9f2f2",
              },
            ]
          }
        >
          <Text />
        </View>
      </View>
    </View>
  </View>
</RCTScrollView>
`;
